### PR TITLE
Support for fixed arrays and structs as function parameters

### PR DESCRIPTION
**Pull request by spencerhaoxiao**
1. Support fixed arrays as function arguments
2. Support the combination of, nested structs and multi-dimensional arrays and all their combinations
3. The predicate on an array argument is to add the relevant predicates on its element type.

### CODE REVIEW

File: src/kontrol/solc_to_k.py [hunk 1]
```diff
@@ -64,6 +64,84 @@ def solc_to_k(
...
+    @staticmethod
+    def from_dict(_input: dict) -> Input:
+        name = _input['name']
+        type = _input['type']
+        if _input.get('components') is not None:
+            if type == 'tuple':
+                components = Input._recurse_comp(name, _input['components'])
+            else:
+                dimension = _find_array_dimensions(type)
+                components = Input._recurse_comp(name + "[]" * dimension, _input['components'])
```
Comment (anvacaru) [inline src/kontrol/solc_to_k.py:83]: The `flake8` that checks if the source code respects the style guides is setup to throw if double quotes are used. You can run locally in the root of the repo `make format; make` to format the source code and run unit tests.

Comment (anvacaru) [reply src/kontrol/solc_to_k.py:83]: That's why the CI is failing here https://github.com/runtimeverification/kontrol/actions/runs/6611613582/job/17955868147?pr=121 .

File: src/kontrol/solc_to_k.py [hunk 2]
```diff
@@ -64,6 +64,84 @@ def solc_to_k(
...
+
+    @staticmethod
+    def _recurse_comp(parent: str, components: dict) -> list[Input]:
+        comps = []
+        for comp in components:
+            _name = comp['name']
+            _type = comp['type']
+            if comp.get('components') is not None:
+                if _type == 'tuple':
+                    new_comps = Input._recurse_comp(parent, _name, comp['components'])
```
Comment (anvacaru) [inline src/kontrol/solc_to_k.py:96]: Input.`_recurse_comp` is defined above with two arguments, yet here the call has three.

**Comment by anvacaru** [comment]
Could you add some unit tests for this in https://github.com/runtimeverification/kontrol/blob/master/src/tests/unit/test_solc_to_k.py?

**Review by anvacaru — APPROVED** [review]

**Comment by palinatolmach** [comment]
@spencerhaoxiao could you please move a Solidity `StructTest` from https://github.com/runtimeverification/kontrol/pull/78 to this PR?https://github.com/runtimeverification/kontrol/blob/c40d8aaf07f1fd341bea5b81067cfbb36df4ee5e/src/tests/integration/test-data/foundry/test/TypeTest.t.sol#L569

**Comment by palinatolmach** [comment]
As discussed with @anvacaru, we might want to remove references to array elements within `#array` and the rule generated for a test function accepting arrays as parameters — at the moment, the later looks like this (note that it doesn't kompile):
```
rule  ( S2KArrayTypeTest . S2KtestZUndstaticZUndarray ( V0_arr : uint256[3] ) => 
   #abiCallData ( "test_static_array" , #array ( #uint256 ( V0_arr_0_ ) , 3:Int , #uint256 ( V0_arr_0_ ) , #uint256 ( V0_arr_1_ ) , #uint256 ( V0_arr_2_ ) , .TypedArgs ) , .TypedArgs ) )
    ensures ( #rangeUInt ( 256 , V0_arr_0_ )
    andBool ( #rangeUInt ( 256 , V0_arr_1_ )
    andBool ( #rangeUInt ( 256 , V0_arr_2_ )
   )))
```

Referencing array elements is impractical for dynamically-sized arrays or statically-sized ones with a large number of elements. This, however, would make generation of range predicates for elements less straightforward, so I'll continue investigation on it.

**Comment by palinatolmach** [comment]
Based on a discussion with @anvacaru and @PetarMax: for static arrays, we should try adding array elements to the left-hand side too.

**Comment by palinatolmach** [comment]
As @anvacaru has suggested, we'll proceed in three different directions: we will
- factor out struct support as implemented in https://github.com/runtimeverification/kontrol/pull/78 back to merge it sooner
- attempt adding support for dynamic arrays by allowing calldata to be symbolic 
- continue working on statically-sized array support as proposed in this PR, i.e., involving references to the elements of the array in the rules generated in `solc_to_k`

**Comment by anvacaru** [comment]
I'm closing this due to staleness.

**Pull request closed without merging**